File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int m;
    static int[] arr;
    static int[] nums;
    static boolean[] visited;
    public static void dfs(int level) {
        StringBuilder sb = new StringBuilder();
        if (level == m) {
            for (int val : arr) {
                sb.append(val + " ");
            }
            System.out.println(sb);
            return;
        }
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                visited[i] = true;
                arr[level] = nums[i];
                dfs(level + 1);
                visited[i] = false;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        arr = new int[m];
        nums = new int[n];
        visited = new boolean[n];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(nums);

        dfs(0);
    }
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.cc
#include <bits/stdc++.h>
using namespace std;

int N, M;
vector<int> nums;
int arr[10];
bool visited[10];

void dfs(int depth) {
	if (depth == M) {
		for (int i = 0; i < M; i++) {
			cout << arr[i] << ' ';
		}
		cout << '\n';
	}
	
	for (int i = 0; i < N; i++) {
		if (!visited[i]) {
			arr[depth] = nums[i];
			visited[i] = true;
			dfs(depth + 1);
			visited[i] = false;
		}
	}
}

int main() {
	cin >> N >> M;
	for (int i = 0; i < N; i++) {
		int temp;
		cin >> temp;
		nums.push_back(temp);
	}
	sort(nums.begin(), nums.end());

	dfs(0);

	return 0;
}